File: app/src/api/scheduling/clients/services/client-service.ts
import { Client } from '@/api/scheduling/clients/data-transfer-objects/client';
import { http } from '@/api/shared/http';
import { ClientCreate } from '@/api/scheduling/clients/data-transfer-objects/client-create';
import { ClientUpdate } from '@/api/scheduling/clients/data-transfer-objects/client-update';

export class ClientService {
    async getAll(): Promise<Client[]> {
        const res = await http.get<any[]>('/client');
        return res.data.map(r => this._map(r));
    }

    async createClient(create: ClientCreate): Promise<Client> {
        const res = await http.post('/client', create);
        return this._map(res.data);
    }

    async updateClient(update: ClientUpdate): Promise<Client> {
        const res = await http.put(`/client/${update.id}`, update);
        return this._map(res.data);
    }

    async deleteClient(client: Client): Promise<void> {
        await http.delete(`client/${client.id}`);
    }

    _map(a: any): Client {
        var c = new Client(a.id, a.name, a.phone, a.email);

        return c;
    }
}

export const clientService = new ClientService();

File: api/src/Domain/Scheduling/Clients/Services/ClientService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.Clients {
    public interface IClientService : IService {
        Task<Client> GetById(Guid id);
        Task<List<Client>> GetByUser(User user);
        Task<Client> Create(ClientCreate create, User user);
        Task Update(Client client, ClientUpdate update);
        Task Delete(Client client);
    }

    [DependencyInjection(RegisterAs = typeof(IClientService))]
    public class ClientService : IClientService {
        IClientRepo repo;
        ClientHasNoAppointmentsSpecification noAppointmentsSpec;

        public ClientService(IClientRepo repo, ClientHasNoAppointmentsSpecification noAppointmentsSpec) {
            this.repo = repo;
            this.noAppointmentsSpec = noAppointmentsSpec;
        }

        public async Task<Client> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<Client>> GetByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<Client> Create(ClientCreate create, User user) {
            var c = Client.Create(
                user.Id,
                create.Name,
                create.Phone,
                create.Email
            );

            await repo.Add(c);
            return c;
        }

        public async Task Update(Client client, ClientUpdate update) {
            client.Name = update.Name;
            client.Phone = update.Phone;
            client.Email = update.Phone;

            await repo.Update(client);
        }

        public async Task Delete(Client client) {
            await noAppointmentsSpec.CheckAndThrow(client);
            await repo.Delete(client);
        }
    }
}